jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java

```diff
@@ -2227,6 +2237,94 @@ public void setNeedClientAuth(boolean needClientAuth)
         {
             super.setNeedClientAuth(needClientAuth);
         }
+
+        /**
+         * @return whether to reject the TLS handshake if the SNI host is present but does not match
+         */
+        @ManagedAttribute("Whether the TLS handshake is rejected if the SNI host is present but does not match")
+        public boolean isRejectUnmatchedSNIHost()
+        {
+            return rejectUnmatchedSNIHost;
+        }
+
+        /**
+         * @param rejectUnmatchedSNIHost whether to reject the TLS handshake if the SNI host is present but does not match
+         */
+        public void setRejectUnmatchedSNIHost(boolean rejectUnmatchedSNIHost)
+        {
+            this.rejectUnmatchedSNIHost = rejectUnmatchedSNIHost;
+        }
+
+        /**
+         * @return the custom function to select certificates based on SNI information
+         */
+        public SNISelector getSNISelector()
+        {
+            return _sniSelector;
+        }
+
+        /**
+         * <p>Sets a custom function to select certificates based on SNI information.</p>
+         *
+         * @param sniSelector the selection function
+         */
+        public void setSNISelector(SNISelector sniSelector)
+        {
+            _sniSelector = sniSelector;
+        }
+
+        @Override
+        public X509 sniSelect(String keyType, Principal[] issuers, SSLSession session, String sniHost, Collection<X509> certificates) throws SSLHandshakeException
+        {
+            if (sniHost == null)
+            {
+                // No SNI, find the default certificate to send, or let the JDK decide.
+                return certificates.stream()
+                    .filter(x509 -> Objects.equals(x509.getAlias(), getCertAlias()))
+                    .findAny()
+                    .orElse(null);
```

Comment:
I think our default implementation should gate this behaviour with a boolean, so if you want to insist on an SNI name you can just set a boolean rather than have to extend or inject a function..... or we come up with some static functions that can be used in a kind of builder pattern to make the function to pass.